src/ndb_core.c

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Roman Gezikov <rgezikov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <string.h>
+
+#include <ch.h>
+#include "ndb_core.h"
+
+/* TODO: Move that to platform specific type mapping layer */
+typedef Mutex ndb_mutex_t;
+typedef CondVar ndb_condvar_t;
+typedef Thread *ndb_thread_t;
+
+/** Write request */
+typedef struct
+{
+    ndb_file_t *file;
+    ndb_element_t element;
+} ndb_write_q_node_t;
+
+/* TODO: Move to platform parameterization module */
+#define NDB_WRITE_QUEUE_SIZE 8
```

Comment:
I haven't thought about this much, but would it be better to add a `nvm_dirty` flag to the metadata struct for each element and then periodically scan through all the elements in the nvm write thread?